File: sort/merge_sort/swift/merge_sort.swift
func mergeSort<T: Comparable>(_ array: [T]) -> [T] {
  guard array.count > 1 else { return array }
  
  let middleIndex = array.count / 2
  
  let leftArray = mergeSort(Array(array[0..<middleIndex]))
  let rightArray = mergeSort(Array(array[middleIndex..<array.count]))
  
  return merge(leftArray, rightArray)
}

func merge<T: Comparable>(_ left: [T], _ right: [T]) -> [T] {
  var leftIndex = 0
  var rightIndex = 0
  
  var orderedArray: [T] = []
  
  while leftIndex < left.count && rightIndex < right.count {
    let leftElement = left[leftIndex]
    let rightElement = right[rightIndex]
    
    if leftElement < rightElement {
      orderedArray.append(leftElement)
      leftIndex += 1
    } else if leftElement > rightElement {
      orderedArray.append(rightElement)
      rightIndex += 1
    } else {
      orderedArray.append(leftElement)
      leftIndex += 1
      orderedArray.append(rightElement)
      rightIndex += 1
    }
  }
  
  while leftIndex < left.count {
    orderedArray.append(left[leftIndex])
    leftIndex += 1
  }
  
  while rightIndex < right.count {
    orderedArray.append(right[rightIndex])
    rightIndex += 1
  }
  
  return orderedArray
}

let array = [7, 2, 6, 3, 9]
print(mergeSort(array))

File: sort/merge_sort/php/merge_sort.php
<?php

/**
 * @param $objects
 * @return array
 */
function mergeSort($objects)
{
    if(count($objects) == 1 ) {
        return $objects;
    }

    $middle = count($objects) / 2;
    $left   = array_slice($objects, 0, $middle);
    $right  = array_slice($objects, $middle);
    $left   = mergeSort($left);
    $right  = mergeSort($right);

    return merge($left, $right);
}

/**
 * @param $left
 * @param $right
 * @return array
 */
function merge($left, $right)
{
    $res = [];

    while (count($left) > 0 && count($right) > 0)
    {
        if($left[0] > $right[0]) {
            $res[] = array_shift($right);
        } else {
            $res[] = array_shift($left);
        }
    }

    return array_merge($res, $left, $right);
}